### PR TITLE
Complete CRUD operations for Collections page

### LINKED ISSUE #246: CRUD operations for Collections page
Complete CRUD operations for Collections page

## Detailed Description
User can now create, update, and delete collections. Page will automatically load existing collections from database.

## Context
- Basic functionality of the bullet journal.
- Data persistence to save state of bullet journal across sessions.

## Possible Implementation
- Mainly revolves around calling transactions on IndexedDB.  
- Make sure to sync across pages.

## Your Environment
* Browser Name and version:
* Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by will-chung**

### CODE REVIEW

File: source/javascript/src/collection-edit.js [hunk 1]
```diff
@@ -72,12 +78,16 @@ function populateTasks (collection) {
...
-  })
+  try {
+    tasks.forEach((task, index) => {
+      const logItem = createLogItem(task)
+      const li = document.createElement('li')
+      li.appendChild(logItem)
+      taskList.appendChild(li)
+    })
+  } catch (err) {
+    console.log('collection has no tasks')
```
Comment (NoahT) [inline source/javascript/src/collection-edit.js]: As a stretch goal, we may want to look into adding a task for error handling (displaying something to the user when there are no collections). @yuzilyu  what do you think?

File: source/javascript/src/collection.js [hunk 2]
```diff
@@ -1,16 +1,81 @@
...
+          console.log('successfully added "' + collectionName + '"')
+        }
+      }
+    }
+  })
+
+  document.querySelector('.collection-area').append(newCollection)
+}
+
+// function getLogInfoAsJSON (cb) {
```
Comment (NoahT) [inline source/javascript/src/collection.js]: Would it be possible to delete this commented block of code? It's my fault for committing it earlier this week (did it for testing purposes before the recent `IndexedDBWrapper.js` changes)

File: source/javascript/src/components/CollectionItem.js [hunk 3]
```diff
@@ -174,6 +247,45 @@ class CollectionItem extends HTMLElement {
...
+        }
+      }
+    }
+  })
+}
+
 customElements.define('collection-item', CollectionItem)
 
-export { CollectionItem }
+export { CollectionItem, wrapper }
```
Comment (NoahT) [inline source/javascript/src/components/CollectionItem.js:299]: did wrapper not work unless it was included in the exports block?

**Review by NoahT — CHANGES_REQUESTED** [review]
Everything I found is listed here, let me know what you think

File: source/javascript/src/collection-edit.js (hunk 1 above)
Comment (will-chung) [reply source/javascript/src/collection-edit.js]: The error handling here isn't actually needed, I added it for debugging and forgot to get rid of it. I think having some watermark that indicates the page is empty would be nice, but definitely a stretch goal.

File: source/javascript/src/collection.js (hunk 2 above)
Comment (will-chung) [reply source/javascript/src/collection.js]: Okay it's gone. That reminds me, I need to clean up my code with comments.

File: source/javascript/src/components/CollectionItem.js (hunk 3 above)
Comment (will-chung) [reply source/javascript/src/components/CollectionItem.js:299]: I noticed we were creating a new wrapper object each time we needed to a make transaction so I decided to make one to share between the files. I don't know if this is best practice though.

File: source/javascript/src/collection-edit.js (hunk 1 above)
Comment (yuzilyu) [reply source/javascript/src/collection-edit.js]: I think it's okay to show an empty page, just make the button to have the text saying "Add your first collection" or something to implicitly tell the user that there isn't any collections yet and they need to add one

File: source/javascript/src/components/CollectionItem.js (hunk 3 above)
Comment (NoahT) [reply source/javascript/src/components/CollectionItem.js:299]: Sounds similar to a singleton design pattern https://www.sitepoint.com/javascript-design-patterns-singleton/; I'd say this is a good idea if we have enough time at the end of the sprint

File: source/javascript/src/collection.js (hunk 2 above)
Comment (NoahT) [reply source/javascript/src/collection.js]: Same here 😀 we'll assign a bunch of 'spring cleaning' towards the end of this sprint if people have the time

**Review by NoahT — APPROVED** [review]
LGTM

**Comment by yuzilyu** [comment]
Alright im merging it

**Pull request merged**